Issue/6775 media refactor

nbradbury opened the pull request:
Fixes #6775 - updates the photo picker and media browser to use similar code for determining whether to allow multi-select, whether to show just images, etc.  This was done primarily by adding a shared [MediaBrowserType](https://github.com/wordpress-mobile/WordPress-Android/blob/28905e701dc01ea4c5b0807d042c31db9624560f/WordPress/src/main/java/org/wordpress/android/ui/media/MediaBrowserType.java) enum, which will make it simpler to tackle #6719.

There should be no change to how either of these features work - this is internal change only.

mzorz commented on WordPress/src/main/java/org/wordpress/android/ui/media/MediaBrowserType.java:
```diff
@@ -0,0 +1,25 @@
+    BROWSER,                 // browse & manage media
+    EDITOR_PICKER,           // select multiple images or videos to insert into a post
+    FEATURED_IMAGE_PICKER,  // select a single image as a featured image
+    GRAVATAR_IMAGE_PICKER;   // select a single image as a gravatar
```
I wonder if we can simplify the types, and just have an enum like this:

```
public enum MediaBrowserType {
    BROWSER,                 // browse & manage media
    MULTIPLE_PICKER,           // select multiple images or videos to insert into a post
    SINGLE_IMAGE_PICKER;        // select a single image as a featured image, a gravatar, etc.

    public boolean isPicker() {
        return this != BROWSER;
    }

    public boolean isBrowser() {
        return this == BROWSER;
    }

    public boolean isSingleImagePicker() {
        return this == SINGLE_IMAGE_PICKER;
    }

    public boolean canMultiselect() {
        return this == MULTIPLE_PICKER;
    }

}

```

(note that the `imagesOnly()` method goes away as it’s the same as `isSingleImagePicker()`)


If we don’t do that above, I think we can still simplify it a bit and at least remove `imagesOnly()` and replace with `isSingleImagePicker()`
(also `imagesOnly` is inconsistent with the form of <verb><state> naming we’re using for the other methods, i.e. is<state> or <can><do_something> like `isPicker` or `canMultiselect`).

mzorz commented:
It's always nice to see inconsistencies going away @nbradbury! these are my notes on the code:

- in this line here https://github.com/wordpress-mobile/WordPress-Android/pull/6785/files#diff-c91873221424e0f9d31b9854c24c385eR346
shouldn’t we use `mBrowserType.canMultiselect()` in the check?


- in https://github.com/wordpress-mobile/WordPress-Android/pull/6785/files#diff-fdc9b5c706ac889526b8df93c55351ceR202 there’s a method `shouldShowTabs` that makes a check for `mBrowserType`. Can we move that method within the `enum` itself as well?


- in this line https://github.com/wordpress-mobile/WordPress-Android/pull/6785/files#diff-fdc9b5c706ac889526b8df93c55351ceR557 there’s a check for `mBrowserType` being of a certain type.  Can we add a `isBrowser()` method to the `enum` and replace every `mBrowserType == MediaBrowserType.BROWSER` check with it?  I feel for example that specific check would be easier to read like this:
```
if (mBrowserType.isBrowser() && !isLongClick
                  || mBrowserType.isPicker() && isLongClick) {		                  
[…]
}
```


- we could replace this line https://github.com/wordpress-mobile/WordPress-Android/pull/6785/files#diff-c91873221424e0f9d31b9854c24c385eR346  with `mBrowserType.canMultiSelect()`

nbradbury commented:
> I wonder if we can simplify the types...

At this stage we could, but once we enable choosing a device photo as a featured image we'll need separate enums (because we'll need to upload the photo for use as a featured image, whereas we don't need to do that for gravatars).

mzorz commented:
> > I wonder if we can simplify the types...

> At this stage we could, but once we enable choosing a device photo as a featured image we'll need separate enums (because we'll need to upload the photo for use as a featured image, whereas we don't need to do that for gravatars).

Understood - let's keep the enum values then 👍

nbradbury commented:
>  I think we can still simplify it a bit and at least remove imagesOnly() and replace with isSingleImagePicker()

Agreed - changed in dc9be66

> Can we add a isBrowser() method to the enum and replace every mBrowserType == MediaBrowserType.BROWSER check with it?

Added in 51abc29

I've addressed your remaining suggestions by adding a `canFilter()` method to `BrowserType` in fb5619a

mzorz commented:
:shipit:
